HW0/P5/P5.txt

```diff
@@ -1 +1,41 @@
 # Your answers here.
+
+# 1. '255 seconds' -- Since they can add the first two numbers together in 1 second and then one more from the list of 254 to the cumulative total
+
+# 2. 'Yes. 34 seconds' -- Split the bags evenly between the 8 employees for 32 bags per employee, which they can add up in 31 seconds in parallel.  Then 3 seconds to merge them (8->4, 4->2, 2->1)
+
+# 3. '8 seconds.' --  2 ^ n = 256 -> n = 8 seconds, or n = log2(256)
+
+# 4. Plot Infinite Employees v. 1 Employee
+import matplotlib.pyplot as plt
+import math
+import numpy as np
+
+def y2(x):
+    new_vals = []
+    for i in x:
+        if i == 0:
+            new_vals = new_vals + [0.]
+        else:
+            new_vals = new_vals + [math.log(i, 2)] 
+    return new_vals
+
+x = np.linspace(0, 256, 256)
+plt.plot(x , x-1, '-b', label='One Employee')
+plt.plot(x , y2(x), '-g', label='Infinite Employees')
+plt.xlabel('Bag Count')
+plt.ylabel('Counting Time')
+plt.legend()
+plt.title('Bag Count v. Counting Time')
+plt.show()
+
+# 5. '15 seconds' -- The formula becomes n = 2log2(256) - 1, since it will take twice as long as before cause we are essentially adding one second to each calculation, minus the first calculation which the employees don't need to talk to anyone for
+
+# 6. One bag at a time:
+#        '259 seconds' -- If you can only hand out 1 bag at a time, 256 seconds to hand out the bags, with each cashier returning to get a second bag.  While cashiers are waiting on other people to receive their bags
```

Comment:
For Q6, what do you think of the following:
If each cashier takes 1 bag, the distribution time will be 256s, communication time will be 8s (8 rounds of communication) and counting time 8s (8 rounds). The total is 272s.